app/controllers/users/products_controller.rb

```diff
@@ -0,0 +1,90 @@
+class Users::ProductsController < ApplicationController
+  before_action :require_login, except: [:index, :show]
+  before_action :current_user_is_owner, only: [:new, :edit, :destroy]
+
+  def index
+    @categories = Category.order(name: :asc)
+
+    scope = Product.where(user_id: params[:user_id], retired: false).order(name: :asc)
+
+    if params[:category].present?
+      scope = scope.joins(:categories).where(categories: {id: params[:category]})
+    end
+
+    @products = scope
+
+    if User.my_account?(current_user, params[:user_id].to_i)
+        @retired_products = Product.where(user_id: params[:user_id], retired: true).order(name: :asc)
+        render action: :my_index
```

Comment:
Unless you'd like these views to be significantly different from one another, you could have a section of the regular index view that only shows up if the requirements are met. You can do this by having a small amount of control flow through interpolated ruby in the view. This can be a partial and need not take up much space.